In [2]:
# import kagglehub
# # Download latest version
# path = kagglehub.dataset_download("promptcloud/amazon-product-dataset-2020")

# print("Path to dataset files:", path)

# xx

In [3]:
import pandas as pd
df=pd.read_csv("amazon_com_ecommerce.csv")
# df.head()

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10002 entries, 0 to 10001
Data columns (total 28 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Uniq Id                10002 non-null  object 
 1   Product Name           10002 non-null  object 
 2   Brand Name             0 non-null      float64
 3   Asin                   0 non-null      float64
 4   Category               9172 non-null   object 
 5   Upc Ean Code           34 non-null     object 
 6   List Price             0 non-null      float64
 7   Selling Price          9895 non-null   object 
 8   Quantity               0 non-null      float64
 9   Model Number           8230 non-null   object 
 10  About Product          9729 non-null   object 
 11  Product Specification  8370 non-null   object 
 12  Technical Details      9212 non-null   object 
 13  Shipping Weight        8864 non-null   object 
 14  Product Dimensions     479 non-null    object 
 15  Im

In [5]:
# delete columns where entire are NAN
df=df.dropna(axis=1,how="all")
df=df.drop(columns=["Upc Ean Code",]) # delete additonal unuseful
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10002 entries, 0 to 10001
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   Uniq Id                10002 non-null  object
 1   Product Name           10002 non-null  object
 2   Category               9172 non-null   object
 3   Selling Price          9895 non-null   object
 4   Model Number           8230 non-null   object
 5   About Product          9729 non-null   object
 6   Product Specification  8370 non-null   object
 7   Technical Details      9212 non-null   object
 8   Shipping Weight        8864 non-null   object
 9   Product Dimensions     479 non-null    object
 10  Image                  10002 non-null  object
 11  Variants               2478 non-null   object
 12  Product Url            10002 non-null  object
 13  Is Amazon Seller       10002 non-null  object
dtypes: object(14)
memory usage: 1.1+ MB


In [6]:
# Define fill map for key columns
fill_values = {
    "Category": "Not available",
    "Selling Price": "Not available",
    "Model Number": "Not available",
    "About Product": "Not available",
    "Product Specification": "Not available",
    "Technical Details": "Not available",
    "Shipping Weight": "Not available",
    "Product Dimensions": "Not available",
    "Variants": "Not available"
}
df.fillna(value=fill_values, inplace=True)
# or do 
# df.fillna("Not available", inplace=True)
# df.info()

In [7]:
# create all-in-text column description for text embedding
# following will be used for similarity search 
def build_text_for_embedding(row):
    parts = []

    parts.append(f"Product Name: {row['Product Name']}")

    if pd.notnull(row['Category']):
        parts.append(f"Category: {row['Category']}")

    # if pd.notnull(row['Selling Price']):
    #     parts.append(f"Price: {row['Selling Price']}")

    # if pd.notnull(row['Model Number']):
    #     parts.append(f"Model: {row['Model Number']}")

    # if pd.notnull(row['About Product']):
    #     parts.append(f"About Product: {row['About Product']}")
        
    # if pd.notnull(row['Product Specification']):
    #     parts.append(f"Specifications: {row['Product Specification']}") 

    # if pd.notnull(row['Technical Details']):
    #     parts.append(f"Technical Details: {row['Technical Details']}")

    # Convert "Y"/NaN into a clean phrase
    # if row.get("Is Amazon Seller") == "Y":
    #     parts.append("Sold by Amazon: Yes")
    # else:
    #     parts.append("Sold by Amazon: No")

    return " | ".join(parts)

df['combined_text'] = df.apply(build_text_for_embedding, axis=1)


In [8]:
df['About Product'].str.len().max()

2614

In [9]:
# craete matadata list, for retrieval information, 
metadata_list = []
for _, row in df.iterrows():
    metadata_list.append({  
        # all other column not included in combine_text
        # "product_id": row["Uniq Id"],
        "image_url": row["Image"],
        "product_url": row["Product Url"],
        "Variants products link": row["Variants"],
        "Shipping Weight":row["Shipping Weight"],
        "Product Dimensions":row["Product Dimensions"],
        "Product Specification":row["Product Specification"],
        "Technical Details":row["Technical Details"],
        "Is Amazon Seller":row["Is Amazon Seller"],
        "Selling Price":row["Selling Price"],
        "Model Number":row["Model Number"],
        "About Product":row["About Product"],
        
        "combined_text": row["combined_text"] # all text contents
    }
                        )


# embedding
- embed text and image with CLIP, then take their average as final
- save embed in local drive for further retrieval


In [10]:
# !pip install transformers pillow torch
from transformers import CLIPProcessor, CLIPModel
from PIL import Image
import requests
import torch
import pandas as pd
import numpy as np
from tqdm import tqdm

# pretrain model
# model = CLIPModel.from_pretrained("openai/clip-vit-base-patch32")
# processor = CLIPProcessor.from_pretrained("openai/clip-vit-base-patch32")

# finetuneed model
model = CLIPModel.from_pretrained("clip-finetuned")
processor = CLIPProcessor.from_pretrained("clip-finetuned")

Using a slow image processor as `use_fast` is unset and a slow processor was saved with this model. `use_fast=True` will be the default behavior in v4.52, even if the model was saved with a slow processor. This will result in minor differences in outputs. You'll still be able to use a slow processor with `use_fast=False`.


In [16]:
# Define Embedding Function
def get_text_embedding(text):
    inputs = processor(text=[text], return_tensors="pt", truncation=True)
    with torch.no_grad():
        return model.get_text_features(**inputs).squeeze().numpy()

def get_image_embedding(url):
    try:
        image = Image.open(requests.get(url, stream=True, timeout=5).raw).convert("RGB")
        inputs = processor(images=image, return_tensors="pt")
        with torch.no_grad():
            return model.get_image_features(**inputs).squeeze().numpy()
    except:
        return None

from tqdm import tqdm
# used for parsing Image column when there are multiple images, just first image to run
def get_first_image_url(image_url_string):
    """Returns the first image URL from a pipe-separated string."""
    if pd.isna(image_url_string):
        return None
    urls = [u.strip() for u in str(image_url_string).split("|") if u.strip()]
    return urls[0] if urls else None



In [12]:
# create embedding DB
# text_embs, image_embs, final_embs = [], [], []

# for _, row in tqdm(df.iterrows(), total=len(df)):
#     text_emb = get_text_embedding(row["combined_text"])
#     # image_emb = get_image_embedding(row["Image"]) #may have multiple images, hence likely fail for some of them
#     first_image_url = get_first_image_url(row["Image"])
#     image_emb = get_image_embedding(first_image_url) if first_image_url else None

#     if image_emb is not None:
#         combined_emb = (text_emb + image_emb) / 2 # use average of text+image embedding
#     else:
#         combined_emb = text_emb  # fallback to text only

#     text_embs.append(text_emb)
#     image_embs.append(image_emb)
#     final_embs.append(combined_emb)

# final_embs = np.vstack(final_embs).astype("float32")  # Shape: (N, 512)
# text_embs = np.vstack(text_embs).astype("float32")
# image_embs = np.vstack(image_embs).astype("float32")

100%|██████████| 10002/10002 [1:51:03<00:00,  1.50it/s] 


In [11]:
import numpy as np
# save embedding - pretrain CLIP
# np.save("final_clip_embeddings.npy", final_embs)
# save embedding - finetune CLIP
# np.save("finetune_clip_embeddings.npy", final_embs)

# load bakc when need
# final_embs = np.load("final_clip_embeddings.npy")
final_embs = np.load("finetune_clip_embeddings.npy")

# faiss search index matching
- IndexFlatL2(d) creates a simple index that compares vectors using L2 (Euclidean) distance

- index.add(final_embs) adds all 10,002 embeddings to the index

- instantly search for the most similar items using vector similarity can tell “which ones are most similar to a query”

In [12]:
import faiss

d = 512  # dimension of embeddings
index = faiss.IndexFlatL2(d)  # use IndexFlatIP for cosine similarity

index.add(final_embs.astype("float32"))

# save this index local if needed
# faiss.write_index(index, "product_index.faiss")


In [13]:
# test query
query_vector = final_embs[0].reshape(1, -1) 
D, I = index.search(query_vector, k=3) 
# I for Indexes of top k most similar vectors
# D for Distances scores

# Get metadata for top match, These are the matched products — ready to send to an LLM, or display in a UI.
top_matches = [metadata_list[i] for i in I[0]]

In [15]:
D

array([[ 0.     , 16.23633, 16.40052]], dtype=float32)

In [16]:
# recall @ function
def recall_at_k(k, index, embeddings):
    hits = 0
    total = len(embeddings)

    for i in range(total):
        query = embeddings[i].reshape(1, -1).astype("float32")
        _, I = index.search(query, k + 1)  # +1 to include the query itself
        neighbors = set(I[0]) - {i}  # remove self

        if i in I[0][:k + 1]:  # check if self is among top-k (excluding self can be optional)
            hits += 1

    return hits / total
for k in [1, 5, 10]:
    recall = recall_at_k(k, index, final_embs)
    print(f"Recall@{k}: {recall:.4f}")


Recall@1: 0.9997
Recall@5: 1.0000
Recall@10: 1.0000


# RAG framework
return both prompt and retrieved items

In [ ]:
# search function
# def rag_text_query(user_query, k=3):
#     query_emb = get_text_embedding(user_query)
#     D, I = index.search(query_emb.reshape(1, -1), k)
#     retrieved_items = [metadata_list[i] for i in I[0]]
#     prompt = build_prompt(user_query, retrieved_items)
#     return prompt, retrieved_items

# def rag_image_query(image_path_or_url, k=5):
#     query_emb = get_image_embedding(image_path_or_url)
#     D, I = index.search(query_emb.reshape(1, -1), k)
#     retrieved_items = [metadata_list[i] for i in I[0]]
#     prompt = build_prompt(user_query, retrieved_items)
#     return prompt, retrieved_items

# Text example
# prompt, retrievd_contend = rag_text_query("Can you show me a picture of the Apple AirPods Pro?")
# print(prompt)

# Image example
# prompt, retrieved = rag_image_query("https://example.com/product.jpg")
# print(prompt)

In [14]:
# LLM have context length limits
def build_prompt(query, top_metadata):
    # get entire metadata into context for prompt
    prompt_parts = []
    for i, item in enumerate(top_metadata):
        section = f"{i+1}."
        for key, value in item.items():
            if pd.notnull(value) and str(value).strip() != "":
                section += f"\n  - {key}: {value}"
        prompt_parts.append(section)

    context = "\n\n".join(prompt_parts)

    prompt = f"""
You are a helpful AI assistant for an e-commerce website. Your job is to answer customer questions based on available product details.

User question:
{query}

Here are some product descriptions that may be relevant:
{context}

Provide an informative and accurate response using the product information above. If multiple items are relevant, mention them. 
If you don't know the answer, say "I’m not sure based on the available product data.
If query request to show picture relate to the product, please return the corresponding iamge URL from the context.
"""
    return prompt.strip()

def get_query_embedding(text_query=None, image_query=None,k=2): # 
    # skip image if it's not provided
    text_emb = get_text_embedding(text_query) if text_query else None
    image_emb = get_image_embedding(image_query) if image_query else None

    # embed query score
    query_emb=None
    if text_emb is not None and image_emb is not None:
        # query_emb=((text_emb + image_emb) / 2).reshape(1, -1)
        query_emb = (0.3 * text_emb + 0.7 * image_emb).reshape(1, -1) # emphasize image emb
    elif text_emb is not None:
        query_emb=text_emb.reshape(1, -1)
    elif image_emb is not None:
        query_emb=image_emb.reshape(1, -1)
    else:
        raise ValueError("Must provide at least text or image.")
    # faiss serach & retrieval
    D, I = index.search(query_emb, k)
    retrieved_items = [metadata_list[i] for i in I[0]]
    
    prompt = build_prompt(text_query, retrieved_items)
    return prompt, retrieved_items



# test cases 1

In [ ]:
image_path=None
text_query="Can you show me some Dinosaur Toys? Also show their pictures"
prompt, items = get_query_embedding(text_query, image_query=image_path,k=5)

from transformers import pipeline
from huggingface_hub import InferenceClient

client = InferenceClient(
    model="mistralai/Mixtral-8x7B-Instruct-v0.1",
    # token="xxxxx"
)

# client = InferenceClient(
#     model="mistralai/Mixtral-8x7B-v0.1",  # This one *does* support text-generation
#     token="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"
# )

response = client.text_generation(
    prompt=prompt,
    max_new_tokens=500,
    temperature=0.05
)


print(response)

HfHubHTTPError: 404 Client Error: Not Found for url: https://router.huggingface.co/together/v1/completions (Request ID: nvuGFax-4yUbBN-947103704eb8e646)

{'message': 'Unable to access model mistralai/Mixtral-8x7B-v0.1. Please visit https://api.together.ai/models to view the list of supported models.', 'type': 'invalid_request_error', 'param': None, 'code': 'model_not_available'}

In [ ]:
items
# retrieved item are poor
# LLM recognize this fact, and give best judgement

# test case 2

In [27]:
text_query="Can you show me the price of DB Longboards? Also show me the product picture"
image_path=None
prompt, items = get_query_embedding(text_query, image_query=image_path,k=5)
items

[{'image_url': 'https://m.media-amazon.com/images/I/A13usaonutL.jpg|https://m.media-amazon.com/images/G/01/Merch/SizeChart/ATVPDKIKX0DER/HOUSEjpg|https://images-na.ssl-images-amazon.com/images/G/01/x-locale/common/transparent-pixel.jpg',
  'product_url': 'https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B07G77DJJW',
  'Variants products link': 'https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B0752XJYNL|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B07537HR1W|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B07535YF4X|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B0752XJYNL|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B07537HQXD|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B0752XK16C|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B075382QRP|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B075386ZN1',
  'Shipping Weight': '4.8 ounces',
  'Product Dimensions': 'Not available',
  'Product Specification': 'Not 

In [29]:
# only 2 item are related to the query, but 2 of them are very well matched
# LLM took right contends and give right answer

# test case 3

In [32]:
text_query="Can you show me some price of some Funko pop? "
image_path=None
prompt, items = get_query_embedding(text_query, image_query=image_path,k=5)
items

[{'image_url': 'https://m.media-amazon.com/images/I/A13usaonutL.jpg|https://m.media-amazon.com/images/G/01/Merch/SizeChart/ATVPDKIKX0DER/HOUSEjpg|https://images-na.ssl-images-amazon.com/images/G/01/x-locale/common/transparent-pixel.jpg',
  'product_url': 'https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B07G77DJJW',
  'Variants products link': 'https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B0752XJYNL|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B07537HR1W|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B07535YF4X|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B0752XJYNL|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B07537HQXD|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B0752XK16C|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B075382QRP|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B075386ZN1',
  'Shipping Weight': '4.8 ounces',
  'Product Dimensions': 'Not available',
  'Product Specification': 'Not 

In [33]:
from transformers import pipeline
from huggingface_hub import InferenceClient

client = InferenceClient(
    model="mistralai/Mixtral-8x7B-Instruct-v0.1",
    token="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"
    # token="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"
    # token="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"
)
response = client.text_generation(
    prompt=prompt,
    max_new_tokens=500,
    temperature=0.05
)
print(response)



Answer:

Here are the prices of some Funko Pop! figures:

1. Funko POP! Movies: Big - Josh - $8.91

Please note that the other products listed do not appear to be Funko Pop! figures.


# test case 4

In [34]:
image_path=None
text_query="What are the features of the Samsung Galaxy S21?"
prompt, items = get_query_embedding(text_query, image_query=image_path,k=5)
items

[{'image_url': 'https://m.media-amazon.com/images/I/A13usaonutL.jpg|https://m.media-amazon.com/images/G/01/Merch/SizeChart/ATVPDKIKX0DER/HOUSEjpg|https://images-na.ssl-images-amazon.com/images/G/01/x-locale/common/transparent-pixel.jpg',
  'product_url': 'https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B07G77DJJW',
  'Variants products link': 'https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B0752XJYNL|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B07537HR1W|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B07535YF4X|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B0752XJYNL|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B07537HQXD|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B0752XK16C|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B075382QRP|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B075386ZN1',
  'Shipping Weight': '4.8 ounces',
  'Product Dimensions': 'Not available',
  'Product Specification': 'Not 

In [ ]:
# LLM realized retrieval were all inrelevent
# give right comments that it doesn know the query answer

# test case 5

In [36]:
image_path="https://images-na.ssl-images-amazon.com/images/I/31io64kmPfL.jpg"
text_query="what's this product and its price"
prompt, items = get_query_embedding(text_query, image_query=image_path,k=5)
items

[{'image_url': 'https://images-na.ssl-images-amazon.com/images/I/31io64kmPfL.jpg|https://images-na.ssl-images-amazon.com/images/I/41ZGO5kBZWL.jpg|https://images-na.ssl-images-amazon.com/images/G/01/x-locale/common/transparent-pixel.jpg',
  'product_url': 'https://www.amazon.com/Funko-Pop-Keychain-Frozen-Elsa/dp/B07V8WGBH4',
  'Variants products link': 'Not available',
  'Shipping Weight': '0.64 ounces',
  'Product Dimensions': 'Not available',
  'Product Specification': 'ProductDimensions:1.5x1.5x1.5inches|ItemWeight:0.16ounces|ShippingWeight:0.64ounces(Viewshippingratesandpolicies)|ASIN:B07V8WGBH4|Itemmodelnumber:40907|Manufacturerrecommendedage:6yearsandup',
  'Technical Details': 'Size:Standard From Frozen 2, Elsa, as a stylized Pocket POP! Keychain from Funko! Figure stands 1.5 inches and comes in a window display box. Check out the other frozen 2 figures from Funko! Collect them all! | 0.64 ounces (View shipping rates and policies)',
  'Is Amazon Seller': 'Y',
  'Selling Price': '

In [ ]:
# retrieval original image, and seems we ask for k=5, 4 others similar thing show up
# LLM can't catch that only the first is the original picture given

# test case 6

In [38]:
image_path="https://images-na.ssl-images-amazon.com/images/I/31An4Vzy6eL.jpg"
text_query="tell me about this product"
prompt, items = get_query_embedding(text_query, image_query=image_path,k=5)
items

[{'image_url': 'https://images-na.ssl-images-amazon.com/images/I/31An4Vzy6eL.jpg|https://images-na.ssl-images-amazon.com/images/I/315AP21SrbL.jpg|https://images-na.ssl-images-amazon.com/images/I/41zUf%2B2OHtL.jpg|https://images-na.ssl-images-amazon.com/images/I/410Pbq%2BxgsL.jpg|https://images-na.ssl-images-amazon.com/images/I/41locO2ICcL.jpg|https://images-na.ssl-images-amazon.com/images/G/01/x-locale/common/transparent-pixel.jpg',
  'product_url': 'https://www.amazon.com/Bandai-Spirits/dp/B07SDC6KZX',
  'Variants products link': 'Not available',
  'Shipping Weight': '3.84 ounces',
  'Product Dimensions': 'Not available',
  'Product Specification': 'ProductDimensions:1.8x1.8x5.9inches|ItemWeight:3.84ounces|ShippingWeight:3.84ounces(Viewshippingratesandpolicies)|ASIN:B07SDC6KZX|Itemmodelnumber:-|Manufacturerrecommendedage:13yearsandup',
  'Technical Details': '3.84 ounces (View shipping rates and policies) | -: - - - - - - - - -',
  'Is Amazon Seller': 'Y',
  'Selling Price': '$13.00',

In [ ]:
# retrieval original image, and seems we ask for k=5, 4 others similar thing show up
# LLM can't catch that only the first is the original picture given

# test case 7

In [40]:
image_path="https://images-na.ssl-images-amazon.com/images/I/31io64kmPfL.jpg"
text_query="give some similar product of this, and their pictures"
prompt, items = get_query_embedding(text_query, image_query=image_path,k=5)
items

[{'image_url': 'https://images-na.ssl-images-amazon.com/images/I/31io64kmPfL.jpg|https://images-na.ssl-images-amazon.com/images/I/41ZGO5kBZWL.jpg|https://images-na.ssl-images-amazon.com/images/G/01/x-locale/common/transparent-pixel.jpg',
  'product_url': 'https://www.amazon.com/Funko-Pop-Keychain-Frozen-Elsa/dp/B07V8WGBH4',
  'Variants products link': 'Not available',
  'Shipping Weight': '0.64 ounces',
  'Product Dimensions': 'Not available',
  'Product Specification': 'ProductDimensions:1.5x1.5x1.5inches|ItemWeight:0.16ounces|ShippingWeight:0.64ounces(Viewshippingratesandpolicies)|ASIN:B07V8WGBH4|Itemmodelnumber:40907|Manufacturerrecommendedage:6yearsandup',
  'Technical Details': 'Size:Standard From Frozen 2, Elsa, as a stylized Pocket POP! Keychain from Funko! Figure stands 1.5 inches and comes in a window display box. Check out the other frozen 2 figures from Funko! Collect them all! | 0.64 ounces (View shipping rates and policies)',
  'Is Amazon Seller': 'Y',
  'Selling Price': '

In [ ]:
# great response

# test case 8

In [42]:
image_path=None
text_query="Can you show me the price of Bandai Spirits Bandai Hobby Mecha Collection Astro Battleship? Also show me the product picture"
prompt, items = get_query_embedding(text_query, image_query=image_path,k=5)
items

[{'image_url': 'https://images-na.ssl-images-amazon.com/images/I/41cz06RaSgL.jpg|https://images-na.ssl-images-amazon.com/images/G/01/x-locale/common/transparent-pixel.jpg',
  'product_url': 'https://www.amazon.com/Banpresto-Mobile-Internal-Structure-Rx-78-2/dp/B082V1PBP5',
  'Variants products link': 'Not available',
  'Shipping Weight': '2 pounds',
  'Product Dimensions': 'Not available',
  'Product Specification': 'ProductDimensions:6x6x8inches|ItemWeight:2pounds|ShippingWeight:2pounds(Viewshippingratesandpolicies)|ASIN:B082V1PBP5|Itemmodelnumber:BP16203|Manufacturerrecommendedage:15yearsandup',
  'Technical Details': 'show up to 2 reviews by default Figure from Banpresto | 2 pounds (View shipping rates and policies)',
  'Is Amazon Seller': 'Y',
  'Selling Price': '$19.99',
  'Model Number': 'BP16203',
  'About Product': 'Make sure this fits by entering your model number. | Base Stand included',
  'combined_text': 'Product Name: Banpresto Mobile Suit Gundam Internal Structure Rx-78-2

# test case 9

In [45]:
image_path=None
text_query="tell me difference between apple iphones and samsung phones"
prompt, items = get_query_embedding(text_query, image_query=image_path,k=5)
items

[{'image_url': 'https://m.media-amazon.com/images/I/A13usaonutL.jpg|https://m.media-amazon.com/images/G/01/Merch/SizeChart/ATVPDKIKX0DER/HOUSEjpg|https://images-na.ssl-images-amazon.com/images/G/01/x-locale/common/transparent-pixel.jpg',
  'product_url': 'https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B07G77DJJW',
  'Variants products link': 'https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B0752XJYNL|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B07537HR1W|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B07535YF4X|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B0752XJYNL|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B07537HQXD|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B0752XK16C|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B075382QRP|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B075386ZN1',
  'Shipping Weight': '4.8 ounces',
  'Product Dimensions': 'Not available',
  'Product Specification': 'Not 

In [47]:
# product doesn't exist, none retrieved
# LLM recognized it, but give contend outside of the databases provided

# test case 10

In [48]:
image_path=None
text_query="what's price range of pokemon toys"
prompt, items = get_query_embedding(text_query, image_query=image_path,k=5)
items

[{'image_url': 'https://m.media-amazon.com/images/I/A13usaonutL.jpg|https://m.media-amazon.com/images/G/01/Merch/SizeChart/ATVPDKIKX0DER/HOUSEjpg|https://images-na.ssl-images-amazon.com/images/G/01/x-locale/common/transparent-pixel.jpg',
  'product_url': 'https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B07G77DJJW',
  'Variants products link': 'https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B0752XJYNL|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B07537HR1W|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B07535YF4X|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B0752XJYNL|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B07537HQXD|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B0752XK16C|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B075382QRP|https://www.amazon.com/Id-Rather-Playing-Board-Games/dp/B075386ZN1',
  'Shipping Weight': '4.8 ounces',
  'Product Dimensions': 'Not available',
  'Product Specification': 'Not 

In [ ]:
# LLM didn't recognize that one of the retrieval was totally inrelevent to query

In [ ]:
# df[df["Product Name"].str.contains("Samsung", case=False, na=False)]

In [50]:
# df["Image"].iloc[250]

In [51]:
# df["Product Name"].head(30)

showing images, needed to deployment only

In [ ]:
# import re
# from IPython.display import display, Image
# from PIL import UnidentifiedImageError

# # Extract all .jpg image URLs from the response
# def extract_all_jpg_urls(response_text):
#     return re.findall(r"https:\/\/[^\s\"']+?\.jpg", response_text, flags=re.IGNORECASE)

# # Display all image URLs found
# def display_images_from_urls(urls):
#     for url in urls:
#         try:
#             display(Image(url=url))
#         except UnidentifiedImageError:
#             print(f"[Unidentified Image] Skipped: {url}")
#         except Exception as e:
#             print(f"[Error] {url} — {e}")

# # Extract and display
# urls = extract_all_jpg_urls(response)
# print(f"Extracted {len(urls)} image URLs.")
# display_images_from_urls(urls)